### ISSUE TITLE
Visualizing the RGB

### ISSUE THREAD

**Issue by eriksandstroem** [youmi-zym/GO-SLAM#13]
Hi,
Thanks a lot for sharing this great project! 

I see that there is a function that can render images from the model: https://github.com/youmi-zym/GO-SLAM/blob/6997c213f35d154b4ff040c5948bcaa0e8fe2ec7/src/image_visualization.py#L25

But I don't see it being used anywhere. Would you be able to give me some hints on where to put this in the code to render the images at the end of trajectory capture?

Cheers,
Erik

**Comment by glaciomarine** [comment]
Can I ask you how to install tiny-cuda-nn

**Comment by youmi-zym** [comment]
Hi Erik,

Yes, I used it to debug, but I accidentally removed it when I cleaned the code. So, to use it, just put the lines below at [here](https://github.com/youmi-zym/GO-SLAM/blob/main/src/mapping.py#L297):

```
# visualize current image prediction
meta = self.video.get_mapping_item(cur_idx, device, decay=1.0)
gt_color, gt_depth, c2w, gt_c2w, mask = meta
self.visualizer.vis(cur_idx, self.local_step, gt_depth, gt_color, c2w, self.mapping_net.to(device))
```

Best,
Youmin

**Comment by youmi-zym** [comment]
Hi @glaciomarine,

you can use this command to install it:

`pip install git+https://github.com/NVlabs/tiny-cuda-nn/#subdirectory=bindings/torch
`

If you meet other issues, please refer to the official guideline.

Best,
Youmin

**Comment by eriksandstroem** [comment]
> Hi Erik,
> 
> Yes, I used it to debug, but I accidentally removed it when I cleaned the code. So, to use it, just put the lines below at [here](https://github.com/youmi-zym/GO-SLAM/blob/main/src/mapping.py#L297):
> 
> ```
> # visualize current image prediction
> meta = self.video.get_mapping_item(cur_idx, device, decay=1.0)
> gt_color, gt_depth, c2w, gt_c2w, mask = meta
> self.visualizer.vis(cur_idx, self.local_step, gt_depth, gt_color, c2w, self.mapping_net.to(device))
> ```
> 
> Best, Youmin

Thank you! 
But won't this simply render the rgb at the current frame, which can be different than at the end of trajectory capture, or you did not note any difference between rendering during runtime and after the trajectory has been fully integrated?

**Comment by youmi-zym** [comment]
Hi,

Okay, I got what you want. 

Firstly, there is a big difference between rendering the images on the fly and the one at the very end. So, I'd advise you to do it after the tracking and mapping are completed.

To be specific, maybe you can put the above lines (with for loop) [here](https://github.com/youmi-zym/GO-SLAM/blob/main/src/mesher.py#L310) to get the rendering results of each keyframe. Moreover, you can also use each frame's pose, rather than each keyframe's pose to get the rendering results of each frame. To be honest, I didn't try this before, so I don't know what would happen.

Best,
Youmin

**Comment by eriksandstroem** [comment]
Hi again,
Are you sure that it is as easy as adding t

> > Hi Erik,
> > Yes, I used it to debug, but I accidentally removed it when I cleaned the code. So, to use it, just put the lines below at [here](https://github.com/youmi-zym/GO-SLAM/blob/main/src/mapping.py#L297):
> > ```
> > # visualize current image prediction
> > meta = self.video.get_mapping_item(cur_idx, device, decay=1.0)
> > gt_color, gt_depth, c2w, gt_c2w, mask = meta
> > self.visualizer.vis(cur_idx, self.local_step, gt_depth, gt_color, c2w, self.mapping_net.to(device))
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Best, Youmin
> 
> Thank you! But won't this simply render the rgb at the current frame, which can be different than at the end of trajectory capture, or did you not note any difference between rendering during runtime and after the trajectory has been fully integrated?

Btw, are you sure it is as easy as adding those lines? I had to remove the self.local_step to make it compatible with the function, but it does not seem to output any files from what I can tell. Must be some minor issue I imagine.

**Comment by eriksandstroem** [comment]
> Hi,
> 
> Okay, I got what you want.
> 
> Firstly, there is a big difference between rendering the images on the fly and the one at the very end. So, I'd advise you to do it after the tracking and mapping are completed.
> 
> To be specific, maybe you can put the above lines (with for loop) [here](https://github.com/youmi-zym/GO-SLAM/blob/main/src/mesher.py#L310) to get the rendering results of each keyframe. Moreover, you can also use each frame's pose, rather than each keyframe's pose to get the rendering results of each frame. To be honest, I didn't try this before, so I don't know what would happen.
> 
> Best, Youmin

Thanks! Yes, I also expect a big difference and that renderings at the end should be better because more information has been integrated. I suppose I can simply load the dataset like is done here: https://github.com/youmi-zym/GO-SLAM/blob/main/run.py#L85
and then loop over it?

**Comment by youmi-zym** [comment]
Hi, 

Yes, correctly! For `self.local_step`, you can set any value you want, maybe just 1. It's just an identifier.

**Comment by eriksandstroem** [comment]
> Hi,
> 
> Yes, correctly! For `self.local_step`, you can set any value you want, maybe just 1. It's just an identifier.

no, it has to be removed, because it is not expected as input to the "vis" function.

**Comment by youmi-zym** [comment]
Alright! Thanks. You're right. Sorry, `cur_idx` is the identifier. To be clear, here is new command:

```
# visualize current image prediction
meta = self.video.get_mapping_item(cur_idx, device, decay=1.0)
gt_color, gt_depth, c2w, gt_c2w, mask = meta
self.visualizer.vis(cur_idx, gt_depth, gt_color, c2w, self.mapping_net.to(device))
```

**Comment by eriksandstroem** [comment]
![image](https://github.com/youmi-zym/GO-SLAM/assets/33764884/566fdab1-d50e-40aa-876a-ceeb3f06597e)

This is the output on scannet scene 59 keyframe 81. I suspect something is wrong. In many of the predicted depth and rgb images they are just constant (black rgb). Is this expected? The mesh that is produced looks like this - this seems to be correct as far as I can tell.
![image](https://github.com/youmi-zym/GO-SLAM/assets/33764884/1800fcb4-9360-46cd-bcf9-be81026fb082)

**Comment by eriksandstroem** [comment]
I implemented the strategy we discussed above by rendering at the end of trajectory capture. I put the code here just below the mesher as I wanted to be able to turn off meshing independently of rendering. I think it should work:
![image](https://github.com/youmi-zym/GO-SLAM/assets/33764884/4b124a2a-420f-4330-a1d4-b95deb54b073)


However, I ran a test setting the `max_frames` variable to 100, and I get similar constant valued output renderings as when I rendered at runtime. Here is an example:
![image](https://github.com/youmi-zym/GO-SLAM/assets/33764884/d8f505f9-05d8-4ce6-8b34-ac2069629c26)


Would greatly appreciate your input on this!
